### PR TITLE
server (ticdc): add timeout for statusServer.

### LINKED ISSUE #5303: Goroutine leak when usign /api/v1/changefeed to create kafka changefeed
### What did you do?

* Create changefeeds using `/api/v1/changefeed`.
* TiCDC fails to create changefeed, because downstream kafka rejects unauthorized TiCDC.

```
[2022/04/26 14:12:26.682 +00:00] [INFO] [http_router.go:148] [/api/v1/changefeeds] [status=500] [method=POST] [path=/api/v1/changefeeds] [query=] [ip=[172.30.104.117](http://172.30.104.117/)] [user-agent=Go-http-client/1.1] [error="[CDC:ErrKafkaNewSaramaProducer]new sarama producer: [CDC:ErrKafkaInvalidConfig]because TiCDC Kafka producer's `request.required.acks` defaults to -1, TiCDC cannot deliver messages when the `replication-factor` is less than `min.insync.replicas`: kafka server: The client is not authorized to send this request type."] [errorVerbose="[CDC:ErrKafkaNewSaramaProducer]new sarama producer: [CDC:ErrKafkaInvalidConfig]because TiCDC Kafka producer's `request.required.acks` defaults to -1, TiCDC cannot deliver messages when the `replication-factor` is less than `min.insync.replicas`: kafka server: The client is not authorized to send this request type.\[ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc/sink/producer/kafka.NewKafkaSaramaProducer\n\tgithub.com/pingcap/tiflow/cdc/sink/producer/kafka/kafka.go:284\ngithub.com/pingcap/tiflow/cdc/sink.newKafkaSaramaSink\n\tgithub.com/pingcap/tiflow/cdc/sink/mq.go:398\ngithub.com/pingcap/tiflow/cdc/sink.init.1.func3\n\tgithub.com/pingcap/tiflow/cdc/sink/sink.go:103\ngithub.com/pingcap/tiflow/cdc/sink.New\n\tgithub.com/pingcap/tiflow/cdc/sink/sink.go:123\ngithub.com/pingcap/tiflow/cdc/sink.Validate\n\tgithub.com/pingcap/tiflow/cdc/sink/sink.go:136\ngithub.com/pingcap/tiflow/cdc/capture.verifyCreateChangefeedConfig\n\tgithub.com/pingcap/tiflow/cdc/capture/http_validator.go:145\ngithub.com/pingcap/tiflow/cdc/capture.(*HTTPHandler).CreateChangefeed\n\tgithub.com/pingcap/tiflow/cdc/capture/http_handler.go:216\ngithub.com/gin-gonic/gin.(*Context).Next\n\tgithub.com/gin-gonic/gin@v1.7.4/context.go:165\ngithub.com/pingcap/tiflow/cdc.errorHandleMiddleware.func1\n\tgithub.com/pingcap/tiflow/cdc/http_router.go:163\ngithub.com/gin-gonic/gin.(*Context).Next\n\tgithub.com/gin-gonic/gin@v1.7.4/context.go:165\ngithub.com/pingcap/tiflow/cdc.timeoutMiddleware.func1\n\tgithub.com/pingcap/tiflow/cdc/http_router.go:126\ngithub.com/gin-gonic/gin.(*Context).Next\n\tgithub.com/gin-gonic/gin@v1.7.4/context.go:165\ngithub.com/pingcap/tiflow/cdc.logMiddleware.func1\n\tgithub.com/pingcap/tiflow/cdc/http_router.go:135\ngithub.com/gin-gonic/gin.(*Context).Next\n\tgithub.com/gin-gonic/gin@v1.7.4/context.go:165\ngithub.com/gin-gonic/gin.(*Engine).handleHTTPRequest\n\tgithub.com/gin-gonic/gin@v1.7.4/gin.go:489\ngithub.com/gin-gonic/gin.(*Engine).ServeHTTP\n\tgithub.com/gin-gonic/gin@v1.7.4/gin.go:445\nnet/http.serverHandler.ServeHTTP\n\tnet/http/server.go:2887\nnet/http.(*conn).serve\n\tnet/http/server.go:1952\nruntime.goexit\n\truntime/asm_amd64.s:1371](http://ngithub.com/pingcap/errors.AddStack/n/tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174/ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs/n/tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164/ngithub.com/pingcap/tiflow/pkg/errors.WrapError/n/tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30/ngithub.com/pingcap/tiflow/cdc/sink/producer/kafka.NewKafkaSaramaProducer/n/tgithub.com/pingcap/tiflow/cdc/sink/producer/kafka/kafka.go:284/ngithub.com/pingcap/tiflow/cdc/sink.newKafkaSaramaSink/n/tgithub.com/pingcap/tiflow/cdc/sink/mq.go:398/ngithub.com/pingcap/tiflow/cdc/sink.init.1.func3/n/tgithub.com/pingcap/tiflow/cdc/sink/sink.go:103/ngithub.com/pingcap/tiflow/cdc/sink.New/n/tgithub.com/pingcap/tiflow/cdc/sink/sink.go:123/ngithub.com/pingcap/tiflow/cdc/sink.Validate/n/tgithub.com/pingcap/tiflow/cdc/sink/sink.go:136/ngithub.com/pingcap/tiflow/cdc/capture.verifyCreateChangefeedConfig/n/tgithub.com/pingcap/tiflow/cdc/capture/http_validator.go:145/ngithub.com/pingcap/tiflow/cdc/capture.(*HTTPHandler).CreateChangefeed/n/tgithub.com/pingcap/tiflow/cdc/capture/http_handler.go:216/ngithub.com/gin-gonic/gin.(*Context).Next/n/tgithub.com/gin-gonic/gin@v1.7.4/context.go:165/ngithub.com/pingcap/tiflow/cdc.errorHandleMiddleware.func1/n/tgithub.com/pingcap/tiflow/cdc/http_router.go:163/ngithub.com/gin-gonic/gin.(*Context).Next/n/tgithub.com/gin-gonic/gin@v1.7.4/context.go:165/ngithub.com/pingcap/tiflow/cdc.timeoutMiddleware.func1/n/tgithub.com/pingcap/tiflow/cdc/http_router.go:126/ngithub.com/gin-gonic/gin.(*Context).Next/n/tgithub.com/gin-gonic/gin@v1.7.4/context.go:165/ngithub.com/pingcap/tiflow/cdc.logMiddleware.func1/n/tgithub.com/pingcap/tiflow/cdc/http_router.go:135/ngithub.com/gin-gonic/gin.(*Context).Next/n/tgithub.com/gin-gonic/gin@v1.7.4/context.go:165/ngithub.com/gin-gonic/gin.(*Engine).handleHTTPRequest/n/tgithub.com/gin-gonic/gin@v1.7.4/gin.go:489/ngithub.com/gin-gonic/gin.(*Engine).ServeHTTP/n/tgithub.com/gin-gonic/gin@v1.7.4/gin.go:445/nnet/http.serverHandler.ServeHTTP/n/tnet/http/server.go:2887/nnet/http.(*conn).serve/n/tnet/http/server.go:1952/nruntime.goexit/n/truntime/asm_amd64.s:1371)"] [cost=812.845849ms]
```

[goroutine-leak.log.zip](https://github.com/pingcap/tiflow/files/8579579/goroutine-leak.log.zip)

### What did you expect to see?

No goroutine leak.

### What did you see instead?

Goroutine leak.

### Versions of the cluster

TiCDC version (execute `cdc version`):

```console
v5.4.0
```

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #5303

### What is changed and how it works?
**Invesgation Result:**
1. The requests sent by the client are not closed properly, resulting in `http.server` connection that cannot be closed and keeps goroutine growing.
2.  `adminClient`  is not closed properly when `NewKafkaSaramaSink` is failed, which can lead to goruotine leaks.

**Waht is changed:**
1. Add `ReadTimeout` and `WriteTimeout` for `statusServer` to limit a connection maxmum alive time.
4. Add `maxConnection` for `listener` of `statusServer` to limit the maxmum concurrent connections.
5. Close  sarama `adminClient` when `NewKafkaSaramaSink` return with error to avoid goroutine leaks.
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

Side effects

Related changes

### Release note 

```release-note
Fix a bug that may causes goroutine leak in http server.
```

### CODE REVIEW

File: cdc/api/open.go [hunk 1]
```diff
@@ -263,7 +263,7 @@ func (h *openAPI) CreateChangefeed(c *gin.Context) {
 		return
 	}
 
-	info, err := verifyCreateChangefeedConfig(c, changefeedConfig, h.capture)
+	info, err := verifyCreateChangefeedConfig(ctx, changefeedConfig, h.capture)
```
Comment (liuzix) [inline cdc/api/open.go:266]: Please add a short note on why we cannot use `c` here

File: cdc/server.go [hunk 2]
```diff
@@ -177,7 +183,8 @@ func (s *Server) startStatusHTTP(lis net.Listener) error {
 	RegisterRoutes(router, s.capture, registry)
 
 	// No need to configure TLS because it is already handled by `s.tcpServer`.
-	s.statusServer = &http.Server{Handler: router}
+	// Add ReadTimeout and WriteTimeout to avoid some abnormal connections never close.
+	s.statusServer = &http.Server{Handler: router, ReadTimeout: 10 * time.Minute, WriteTimeout: 10 * time.Minute}
```
Comment (liuzix) [inline cdc/server.go:187]: These numbers should be defined as consts.

File: cdc/sink/mq/mq.go [hunk 3]
```diff
@@ -363,6 +364,16 @@ func NewKafkaSaramaSink(ctx context.Context, sinkURI *url.URL,
 	filter *filter.Filter, replicaConfig *config.ReplicaConfig,
 	opts map[string]string, errCh chan error,
 ) (*mqSink, error) {
+	// we must close adminClient when this func return cause by an error
+	// otherwise the adminClient will never be closed and lead to an goroutine leak
+	var err error
+	var adminClient ckafka.ClusterAdminClient
+	defer func() {
```
Comment (liuzix) [inline cdc/sink/mq/mq.go:371]: Why not move the `defer` to after `kafka.NewAdminClientImpl`?

File: cdc/server.go [hunk 4]
```diff
@@ -166,8 +170,10 @@ func (s *Server) Run(ctx context.Context) error {
 
 // startStatusHTTP starts the HTTP server.
 // `lis` is a listener that gives us plain-text HTTP requests.
-// TODO can we decouple the HTTP server from the capture server?
+// TODO: can we decouple the HTTP server from the capture server?
 func (s *Server) startStatusHTTP(lis net.Listener) error {
+	// limit the max concurrent conections of this statusServer.
+	lis = netutil.LimitListener(lis, maxHTTPConnection)
```
Comment (overvenus) [inline cdc/server.go:176]: What happens when it reaches limit?

Comment (liuzix) [reply cdc/server.go:176]: The TCP connections are queued up and no new goroutines will be created until a connection is processed.

File: cdc/server.go [hunk 5]
```diff
@@ -30,8 +30,10 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"go.etcd.io/etcd/client/pkg/v3/logutil"
 	clientv3 "go.etcd.io/etcd/client/v3"
+
```
Comment (hi-rustin) [inline cdc/server.go:33]:
```suggestion
```

**Review by hi-rustin — COMMENTED** [review]
rest LGTM

**Review by liuzix — APPROVED** [review]

**Review by CharlesCheung96 — APPROVED** [review]

File: cdc/server.go (hunk 4 above)
Comment (overvenus) [reply cdc/server.go:181]: Are queued connections affected by `ReadTimeout` and `WriteTimeout`? Does TiCDC close connections actively if they are timeout?

Comment (overvenus) [reply cdc/server.go:181]: Can TiCDC log closed connections?

**Pull request merged**